api/kroxylicious-filter-api/src/main/java/io/kroxylicious/proxy/filter/FilterInvokers.java

```diff
@@ -0,0 +1,120 @@
+     * @param filter the Filter to create an invoker for
+     * @return the invoker
+     */
+    public static FilterInvoker from(KrpcFilter filter) {
+        boolean isResponseFilter = filter instanceof ResponseFilter;
+        boolean isRequestFilter = filter instanceof RequestFilter;
+        boolean isAnySpecificFilterInterface = SpecificFilterInvoker.implementsAnySpecificFilterInterface(filter);
+        if (isAnySpecificFilterInterface && (isRequestFilter || isResponseFilter)) {
+            throw new IllegalArgumentException("Cannot mix specific message filter interfaces and all message filter interfaces");
+        }
+        if (!isRequestFilter && !isResponseFilter && !isAnySpecificFilterInterface) {
+            throw new IllegalArgumentException("KrpcFilter must implement ResponseFilter, RequestFilter or any combination of specific message Filter interfaces");
+        }
+        if (isResponseFilter && isRequestFilter) {
+            return requestResponseInvoker((RequestFilter) filter, (ResponseFilter) filter);
+        }
+        else if (isRequestFilter) {
+            return requestInvoker((RequestFilter) filter);
+        }
+        else if (isResponseFilter) {
+            return responseInvoker((ResponseFilter) filter);
+        }
+        else {
+            return new SpecificFilterInvoker(filter);
+        }
+    }
+
+    private static FilterInvoker requestInvoker(RequestFilter filter) {
+        return new RequestFilterInvoker(filter);
+    }
+
+    private static FilterInvoker responseInvoker(ResponseFilter filter) {
+        return new ResponseFilterInvoker(filter);
+    }
+
+    private static FilterInvoker requestResponseInvoker(RequestFilter requestFilter, ResponseFilter responseFilter) {
+        return new RequestResponseInvoker(requestFilter, responseFilter);
+    }
+
+    private record RequestFilterInvoker(RequestFilter filter) implements FilterInvoker {
+
+    @Override
+    public void onRequest(ApiKeys apiKey, RequestHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
+        filter.onRequest(apiKey, header, body, filterContext);
+    }
+
+    @Override
+    public boolean shouldHandleRequest(ApiKeys apiKey, short apiVersion) {
+        return filter.shouldHandleRequest(apiKey, apiVersion);
+    }
+
+    }
+
+    private record ResponseFilterInvoker(ResponseFilter filter) implements FilterInvoker {
+
+    @Override
+    public void onResponse(ApiKeys apiKey, ResponseHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
+        filter.onResponse(apiKey, header, body, filterContext);
+    }
+
+    @Override
+    public boolean shouldHandleResponse(ApiKeys apiKey, short apiVersion) {
+        return filter.shouldHandleResponse(apiKey, apiVersion);
+    }
+
+    }
+
+    private record RequestResponseInvoker(RequestFilter requestFilter, ResponseFilter responseFilter) implements FilterInvoker {
```

Comment:
the record formatting is ugly but I've gone with the maven plugin output

Reply:
We should find a way of collecting formatting issues or just incrementally update the plug-in rules where appropriate (probably a separate PR)